File: PartitionEqualSubsetSum.py
class Solution:
    def canPartition(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        target = sum(nums)
        if target % 2 == 1:
            return False
        target //= 2

        dp = [False] * (target + 1)
        dp[0] = True
        for num in nums:
            for i in range(target, num - 1, -1):
                dp[i] = dp[i] | dp[i - num]
        return dp[target]


if __name__ == '__main__':
    nums = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.canPartition(nums))

File: PartitionEqualSubsetSum.cpp
#include<iostream>
#include<vector>
#include<algorithm>

using namespace std;

class Solution {
public:
    bool canPartition(vector<int>& nums) {
        
        int sum = 0;
        for(int num : nums) sum += num;
        
        if(sum % 2) return false;
        
        sum /= 2;
        
        int n = nums.size();
        
        vector<bool> dp(sum + 1);
        dp[0] = true;
        
        for(int num : nums) {
        	for(int i = sum; i >= num; i--) dp[i] = dp[i] | dp[i - num];
		}
		
		return dp[sum];
    }
};

int main() {
	
	int n;
	cin>>n;
	
	vector<int> nums;
	for(int i = 0; i < n; i++) {
		int num;
		cin>>num;
		nums.push_back(num);
	}
	
	Solution *solution = new Solution();
	solution->canPartition(nums) ? cout<<"true" : cout<<"false";
	
	return 0;
	
}
